### PR TITLE
Docker: add dockerfile and README.md

### PR DESCRIPTION
**Pull request by 1570005763**
1. Introduce CoCo/guest-components as a submodule.
2. Offer two types of AA images for gRPC and ttrpc interface.
3. Provide a guide for image building and running in readme file.

### CODE REVIEW

File: Dockerfile.ttrpc [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+# Copy attestation-agent Binary
+COPY --from=builder /usr/src/guest-components/target/x86_64-unknown-linux-gnu/release/attestation-agent /usr/local/bin/attestation-agent
+
+# Set Environment Parameters
+ENV AA_KBC_PARAMS=${AA_KBC_PARAMS}
+ENV RUST_LOG=${RUST_LOG}
+
+# Start attestation-agent with Default Unix Sock (/run/confidential-containers/attestation-agent/attestation-agent.sock)
+CMD [ "attestation-agent" ]
```
Comment (Xynnn007) [inline Dockerfile.ttrpc:54]: 在这块加一个VOLUME，并指定这个sock文件，搜一下dockerfile文档里的VOLUME写法

**Review by Xynnn007 — COMMENTED** [review]
good. 另外可以看一下如何在github action里增加镜像发布功能，这样能把AA的镜像直接通过github action流水线构建发布到package里

参考https://github.com/confidential-containers/trustee/blob/main/.github/workflows/kbs-build-and-push.yaml

File: Dockerfile.ttrpc (hunk 1 above)
Comment (1570005763) [reply Dockerfile.ttrpc:54]: Great suggestions. And I should be opening other PRs to implement CI/CD and image release functionalities.

File: Dockerfile.ttrpc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+FROM ubuntu:22.04
+
+LABEL org.opencontainers.image.source="https://github.com/inclavare-containers/attestation-agent"
+
+# Use Default KBC: cc_kbc, and Local KBS with Address of http://127.0.0.1:8080.
+ARG AA_KBC_PARAMS=cc_kbc::http://127.0.0.1:8085
+ARG RUST_LOG=debug
+
+# Install Runtime Dependencies
+RUN apt-get update && apt-get install -y protobuf-compiler
```
Comment (Xynnn007) [inline Dockerfile.ttrpc]: 我们需要这个吗？

File: Dockerfile.ttrpc [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+# Install TDX Runtime Dependencies
+RUN apt-get update && apt-get install curl gnupg openssl -y && \
+    rm -rf /var/lib/apt/lists/{apt,dpkg,cache,log} /tmp/* /var/tmp/*
+
+RUN curl -L https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key | tee intel-sgx-deb.key | apt-key add - && \
+    echo 'deb [arch=amd64] https://download.01.org/intel-sgx/sgx_repo/ubuntu focal main' | tee /etc/apt/sources.list.d/intel-sgx.list && \
+    apt-get update && apt-get install -y libtdx-attest-dev libsgx-dcap-quote-verify
```
Comment (Xynnn007) [inline Dockerfile.ttrpc]: 我建议把这两大坨合成一个RUN语句，然后最后一并执行rm -rf /var/lib/apt/lists/{apt,dpkg,cache,log} /tmp/* /var/tmp/*

File: README.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Attestation Agent (AA) is a service function set for attestation procedure in Confidential Containers. It provides several kinds of service APIs related to attestation.
```
Comment (jiazhang0) [inline README.md]: 改成：`(aka AA)`

Comment (jiazhang0) [reply README.md]: s/service function set/service/

Comment (jiazhang0) [reply README.md]: s/for attestation/for the attestation/

Comment (jiazhang0) [reply README.md]: s/Confidential Containers/Confidential Computing/

File: README.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+
+For more details, please visit [Attestation Agent](https://github.com/confidential-containers/guest-components/blob/main/attestation-agent/README.md)
+
+This repository packages the Attestation Agent (AA) into an easy-to-use image.
```
Comment (jiazhang0) [inline README.md]: s/Attestation Agent (AA)/AA/

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: README.md [hunk 6]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
 
 For more details, please visit [Attestation Agent](https://github.com/confidential-containers/guest-components/blob/main/attestation-agent/README.md)
 
-This repository packages the Attestation Agent (AA) into an easy-to-use image.
\ No newline at end of file
+This repository packages AA into an easy-to-use image.
```
Comment (jiazhang0) [inline README.md:3]: Move `For more details, please visit [Attestation Agent](https://github.com/confidential-containers/guest-components/blob/main/attestation-agent/README.md)` behind `This repository packages AA into an easy-to-use image.`

File: README.md [hunk 7]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+
+# Build Attestation Agent (AA) Image
```
Comment (jiazhang0) [inline README.md]: 不用每次都这样写了。你最前面写清楚AA是缩写，后面都用缩写AA就行了。

File: README.md [hunk 8]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+
+AA offers 2 kinds of interface, gRPC and ttrpc. Please choose corresponding dockerfile to build your AA image.
```
Comment (jiazhang0) [inline README.md]: s/dockerfile/Dockerfile/

File: README.md [hunk 9]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+
+## Prerequisites
+
+- A Docker-enabled system
```
Comment (jiazhang0) [inline README.md]: s/Docker-enabled system/Ensure Docker installed on your system/

File: README.md [hunk 10]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+
+## Build Image
+
+```
```
Comment (jiazhang0) [inline README.md]: 在```后面添加`shell`

File: README.md [hunk 11]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+# download code
+git clone --recurse-submodules git@github.com:1570005763/attestation-agent-image.git && cd attestation-agent-image
```
Comment (jiazhang0) [inline README.md]: 这行超过80个字符了，改成：
```
git clone --recurse-submodules \
    git@github.com:1570005763/attestation-agent-image.git && \
    cd attestation-agent-image
```

File: README.md [hunk 12]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+docker build -f ./Dockerfile.ttrpc -t attestation-agent:ttrpc .
+```
+
+# Run Attestation Agent (AA) image with Docker
+
+gRPC image and ttrpc image run in a slightly different way.
+
+## Prerequisites
+
+- A Docker-enabled system
```
Comment (jiazhang0) [inline README.md]: 改下

File: README.md [hunk 13]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+- Already have AA image
```
Comment (jiazhang0) [inline README.md]: 这啥表述啊？改成Installed AA image

File: README.md [hunk 14]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+- TEE-supported CPU, including:
+    - Intel TDX
+    - Intel SGX DCAP
```
Comment (jiazhang0) [inline README.md]: 去掉DCAP

File: README.md [hunk 15]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+    - AMD SEV-SNP
+    - Azure SEV-SNP CVM
+    - Azure TDX CVM
+    - Arm Confidential Compute Architecture (CCA)
+
+## Config
```
Comment (jiazhang0) [inline README.md]: Configuration

File: README.md [hunk 16]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+
+You can modify the configuration of AA by setting the following environment variables when starting the container.
+
+### AA_KBC_PARAMS
+
+- Usage： specify the KBC type and KBS address.
```
Comment (jiazhang0) [inline README.md]: 冒号后面多了个空格

File: README.md [hunk 17]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+- Optional Value: AA_KBC_PARAMS=cc_kbc::{KBS_address}
+- Default Value: AA_KBC_PARAMS=cc_kbc::http://127.0.0.1:8085
+
+### RUST_LOG
+
+- Usage： determine log level.
```
Comment (jiazhang0) [inline README.md]: ditto

File: README.md [hunk 18]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+## gRPC interface
+
+Run image in the background.
+```
+docker run -d \
+    -p 50002:50002 \
+    --name attestation-agent-grpc attestation-agent:grpc
+```
+
+Or run image in interactive mode.
```
Comment (jiazhang0) [inline README.md]: 这段就没啥用，属于教不懂docker的人的步骤？需要在这里教吗？删除

File: README.md [hunk 19]
```diff
@@ -2,4 +2,93 @@ Attestation Agent (AA) is a service function set for attestation procedure in Co
...
+## ttrpc interface
+
+Run image in the background.
+```
+docker run -d \
+    -v /run/confidential-containers/attestation-agent/:/run/confidential-containers/attestation-agent/ \
+    --name attestation-agent-ttrpc attestation-agent:ttrpc
+```
+
+Or run image in interactive mode.
```
Comment (jiazhang0) [inline README.md]: ditto

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: Dockerfile.grpc [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+FROM rust:latest as builder
+
+WORKDIR /usr/src/guest-components
+COPY ./coco-guest-components .
+
+# Install Build Dependencies
+RUN apt-get update && apt-get install -y clang libtss2-dev protobuf-compiler
+
+# Install SGX && TDX Build Dependencies
+RUN curl -L https://download.01.org/intel-sgx/sgx_repo/ubuntu/intel-sgx-deb.key | tee intel-sgx-deb.key | apt-key add - && \
```
Comment (jiazhang0) [inline Dockerfile.grpc]: 这种超过80行的都改成\换行加4空格缩进，否则可读性太差

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: Dockerfile.ttrpc (hunk 1 above)
Comment (1570005763) [reply Dockerfile.ttrpc:54]: VOLUME added.

File: Dockerfile.ttrpc (hunk 2 above)
Comment (1570005763) [reply Dockerfile.ttrpc]: Removed.

File: Dockerfile.ttrpc (hunk 3 above)
Comment (1570005763) [reply Dockerfile.ttrpc]: Verified.

File: Dockerfile.grpc (hunk 20 above)
Comment (1570005763) [reply Dockerfile.grpc]: 这里我增加了换行。不过这里超链接比较长，可能没法严格限制在80字符以内。

**Review by jiazhang0 — APPROVED** [review]

**Pull request merged**